Register Fully Kiosk services regardless of setup result

mheath opened the pull request:
If HA is unable to connect to Fully Kiosk, the services don't get registered. This can cause repair to create notifications saying that the 'fully_kiosk.load_url' service is unknown.

Fixes #85444

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

mheath committed:
Register services at integration level

If HA is unable to connect to Fully Kiosk, the services don't get
registered. This can cause repair to create notifications saying
that the 'fully_kiosk.load_url' service is unknown.

Fixes #85444

MartinHjelmare reviewed (changes requested):
NabuCasa core meeting has discussed this PR and our general approach to registering services. We think that we should register integration level services from `async_setup` which runs before `async_setup_entry` and regardless if a config entry is setup. This will give better feedback to the user, ie why the service call failed instead of a missing service, and also allows users to create automations for integrations regardless of the status of config entries.

Please move the service registration to the function `async_setup` and make sure that the service handler checks that the config entry is present and setup before calling config entry specific client APIs.

mheath commented:
I implemented the recommendation made by @MartinHjelmare.

MartinHjelmare reviewed (changes requested):
Please also check that the config entry is setup, ie the state of the config entry. Otherwise the config entry specific items will be missing from `hass.data`.

https://github.com/home-assistant/core/blob/3e5e937541dd368f3d70c7557ceaa7a07fd284e2/homeassistant/components/fully_kiosk/services.py#L50-L55

mheath commented:
I added the check you requested @MartinHjelmare.

MartinHjelmare commented:
Please don't squash or amend commits after review has started to make it easier for readers to track changes.

MartinHjelmare commented on homeassistant/components/fully_kiosk/services.py:
```diff
@@ -52,6 +52,8 @@ async def execute_service(
                         continue
                     if entry.domain != DOMAIN:
                         continue
+                    if DOMAIN not in hass.data or key not in hass.data[DOMAIN]:
```
I think we should check the state of the config entry instead. It should be `ConfigEntryState.LOADED`. We can combine all the checks with `or` as the result should be the same.

mheath committed:
Validate config entry is loaded

MartinHjelmare reviewed (approved):
Thanks!

MartinHjelmare reviewed (changes requested):
I noticed there are tests for the services. Then we should add a test for this case too. Please test calling the service when the config entry is not loaded. Eg, set up the config entry first, then unload it, then call the service and assert that there was no call to the library.

https://github.com/home-assistant/core/blob/rc/tests/components/fully_kiosk/test_services.py

MartinHjelmare commented on homeassistant/components/fully_kiosk/services.py:
```diff
@@ -48,9 +49,11 @@ async def execute_service(
+                        or entry.domain != DOMAIN
+                        or entry.state != ConfigEntryState.LOADED
+                    ):
                         continue
```
If we don't find any matching coordinators, ie config entries, for the device at all, we should raise `HomeAssistantError` at the end of the function. This will notify the user that something went wrong with the service call.

mheath replied:
Thanks, I was going to ask about that. Doing nothing didn't seem like the right thing to do but it looks like a lot of integrations fail silently.

mheath replied:
@MartinHjelmare I have refactored how services get invoked by the Fully Kiosk Browser integration so that invalid device ids provided by the user will result in a `HomeAssistantError`. If the device's config entry is not loaded, a warning will be logged. I have added tests that verify the `HomeAssistantError` is getting raised as well as ensuring that an unloaded config entry will not be used.

Any feedback would be appreciated.

mheath committed:
Refactor service invocation

Raises `HomeAssistantError` when the user provides an device id that is
not in the device registry or a device that is not a Fully Kiosk
device. If the device's config entry is not loaded, a warning is
logged.

mheath committed:
Merge branch 'dev' into fully_kiosk_unknown_service

mheath commented:
@MartinHjelmare Any chance you could take another look at this change? There's another service I would like to add to the FK integration but I want to get this change resolved first. Thanks.

MartinHjelmare commented on homeassistant/components/fully_kiosk/services.py:
```diff
@@ -20,55 +18,51 @@
+                )
+        coordinators = list[FullyKioskDataUpdateCoordinator]()
+        for config_entry in config_entries:
+            if config_entry.state == ConfigEntryState.LOADED:
+                coordinators.append(hass.data[DOMAIN][config_entry.entry_id])
+            else:
+                LOGGER.warning("Config for %s is not loaded", config_entry.title)
```
I think we should raise HomeAssistantError here instead. This integration should normally only have one config entry per coordinator and device so I don't think we need to allow other coordinators to be collected if a config entry is not loaded.

```suggestion
            if config_entry.state != ConfigEntryState.LOADED:
                raise HomeAssistantError(f"{config_entry.title} is not loaded")
            coordinators.append(hass.data[DOMAIN][config_entry.entry_id])
```

Commit:
Update homeassistant/components/fully_kiosk/services.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

mheath committed:
Assert HomeAssistantError when integration unloaded

mheath committed:
Remove unused import

mheath commented:
@MartinHjelmare I brought in your suggested change and updated the tests. The `hassfest` check if failing but that looks unrelated to this change.

mheath committed:
Merge branch 'dev' into fully_kiosk_unknown_service

mheath committed:
Set CONFIG_SCHEMA

mheath commented:
Nevermind. I fixed the `hassfest` task failure.

MartinHjelmare commented on homeassistant/components/fully_kiosk/__init__.py:
```diff
@@ -16,6 +18,16 @@
     Platform.SWITCH,
 ]
 
+CONFIG_SCHEMA = cv.empty_config_schema(DOMAIN)
```
```suggestion
CONFIG_SCHEMA = cv.config_entry_only_config_schema(DOMAIN)
```

MartinHjelmare commented on homeassistant/components/fully_kiosk/services.py:
```diff
@@ -1,74 +1,66 @@
+                    entry = hass.config_entries.async_get_entry(entry_id)
+                    if entry and entry.domain == DOMAIN:
+                        device_entries.append(entry)
+                if len(device_entries) == 0:
```
```suggestion
                if not device_entries:
```

MartinHjelmare commented on homeassistant/components/fully_kiosk/services.py:
```diff
@@ -1,74 +1,66 @@
+                        device_entries.append(entry)
+                if len(device_entries) == 0:
+                    raise HomeAssistantError(
+                        f"Device '{target}' is not a {DOMAIN} device"
```
Please add a test for this case.

Commit:
Update homeassistant/components/fully_kiosk/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

mheath committed:
Add test for non fkb devices targets in service calls

mheath committed:
Merge branch 'dev' into fully_kiosk_unknown_service

mheath commented:
I merged in your suggestions and added the test.

Remember back when this PR was just a simple 2 line change... :)

MartinHjelmare commented on tests/components/fully_kiosk/test_services.py:
```diff
@@ -103,3 +103,39 @@ async def test_service_bad_device_id(
+    other_config_id = "555"
+    await hass.config_entries.async_add(
+        MockConfigEntry(
+            title="Not Fully Kiosk", domain=other_domain, unique_id=other_config_id
```
```suggestion
            title="Not Fully Kiosk", domain=other_domain, entry_id=other_config_id
```

MartinHjelmare committed:
Apply suggestions from code review

MartinHjelmare reviewed (approved):
Thanks!
